File: Code_C#.NET/MP07_ChuyenRGBsangHSI.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using Emgu.CV;
using Emgu.CV.Structure;
using Emgu.Util;

namespace MP02_TachAnhMauRGB_EmguCV
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
            //Tao bien chua duong dan luu hinh cua minh
            Image<Bgr, byte> hinhgoc = new Image<Bgr, byte>(@"H:\Code\Photo\lena_color.jpg");

            //Hien thi hinh goc trong picBox_Hinhgoc da tao
            imageBox_Hinhgoc.Image = hinhgoc;
            Image<Bgr, byte> Hue = new Image<Bgr, byte>(hinhgoc.Width, hinhgoc.Height);
            Image<Bgr, byte> Saturation = new Image<Bgr, byte>(hinhgoc.Width, hinhgoc.Height);
            Image<Bgr, byte> Intensity = new Image<Bgr, byte>(hinhgoc.Width, hinhgoc.Height);
            Image<Bgr, byte> HSI = new Image<Bgr, byte>(hinhgoc.Width, hinhgoc.Height);

            //Moi hinh la mot ma tran 2 chieu nen se dung 2 vong lap for
            //de doc het cac diem anh co trong hinh
            for (int x = 0; x < hinhgoc.Width; x++)
            {
                for (int y = 0; y < hinhgoc.Height; y++)
                {
                    //doc gia tri pixel tai diem anh co vi tri (x,y)
                    Bgr pixelValue = hinhgoc[y, x];

                    //Tao ra 3 bien R,G,B de lay gia tri RED, GREEN, BLUE tai tuong pixel
                    //Tao bien K de gan gia tri nho nhat cua R,G,B vao
                    //Dunng kieu du lieu double vi trong qua trinh tinh toan cac gia tri H,S,I
                    //co tra ve gia tri so thuc
                    double R = pixelValue.Red;
                    double G = pixelValue.Green;
                    double B = pixelValue.Blue;        

                    //Tinh goc theta dua tren cong thuc
                    double t1 = ((R - G) + (R - B)) / 2;
                    double t2 = Math.Sqrt((R - G) * (R - G) + (R - B)*(G - B));
                    double theta = Math.Acos(t1 / t2);

                    //Tinh gia tri H
                    double H = 0;
                    if (B <= G)
                        H = theta;
                    else
                        H = 2*Math.PI - theta; //Gia tri goc hien tai dang la Radian

                    H = H * 180 / Math.PI; //Chuyen ve gia tri Do

                    //Tinh gia tri S
                    double S = 1 - 3 * Math.Min(R, Math.Min(G, B)) / (R + G + B);

                    //Tinh gia tri I
                    double I = (R + G + B) / 3;

                    //Gan cac gia tri vao

                    Hue.Data[y, x, 2] = (byte)H;
                    Hue.Data[y, x, 1] = (byte)H;
                    Hue.Data[y, x, 0] = (byte)H;

                    
                    Saturation.Data[y, x, 2] = (byte)(S*255);
                    Saturation.Data[y, x, 1] = (byte)(S*255);
                    Saturation.Data[y, x, 0] = (byte)(S*255);

                    
                    Intensity.Data[y, x, 2] = (byte)I;
                    Intensity.Data[y, x, 1] = (byte)I;
                    Intensity.Data[y, x, 0] = (byte)I;

                    
                    HSI.Data[y, x, 2] = (byte)H;
                    HSI.Data[y, x, 1] = (byte)(S*255);
                    HSI.Data[y, x, 0] = (byte)I;


                }
            }
            //Hien thi cac kenh mau len imageBox
            imageBox_H.Image = Hue;
            imageBox_S.Image = Saturation;
            imageBox_I.Image = Intensity;
            imageBox_HSI.Image = HSI;
        }

        private void Form1_Load(object sender, EventArgs e)
        {

        }

        private void imageBox_Hinhgoc_Click(object sender, EventArgs e)
        {

        }

        private void label2_Click(object sender, EventArgs e)
        {

        }
    }
}

File: Code_python/MP07_ChuyenRGBSangHSI.py
import cv2 #Xu li anh
from PIL import Image #Thu vien xu li anh ho tro nhieu loai anh
import numpy as np
import matplotlib.pyplot as plt #Ve bieu do histogram
import math
#Tao duong dan
filehinh = r'lena_color.jpg'
#Đọc ảnh màu dùng thư viện OpenCV
img = cv2.imread(filehinh, cv2.IMREAD_COLOR)
#Đọc ảnh màu dùng thư viện PIL

imgPIL = Image.open(filehinh)
#Lay kich thuoc cua anh
height, width, channel = img.shape

#Khai bao 3 bien de chua gia tri 4 kenh C,M,Y,K
Hue = Image.new(imgPIL.mode, imgPIL.size)
Saturation = Image.new(imgPIL.mode, imgPIL.size)
Intensity = Image.new(imgPIL.mode, imgPIL.size)
HSI = Image.new(imgPIL.mode, imgPIL.size)

#Dung vong for de doc het cac diem anh
for x in range(width):
    for y in range(height):

        #Lay gia tri R,G,B cua tung pixel va gan vao cac bien R,G,B
        R, G, B = imgPIL.getpixel((x,y))      

        t1 = ((R - G) + (R - B)) / 2
        t2 = math.sqrt((R - G) * (R - G) + (R - B)*(G - B))
        theta = math.acos(t1 / t2)

        #Tinh gia tri H
        H = 0
        if (B<= G):
            H = theta
        else :
            H = math.pi*2 - theta
        H = np.uint8(H*180/math.pi)

        #Tinh gia tri S
        Min = min(R,G,B)
        S = np.uint8((1 - 3*Min/(R+G+B))*255)

        #Tinh gia tri I
        I = np.uint8((R + G + B)/3)
        
        Hue.putpixel((x,y),(H,H,H))

        Saturation.putpixel((x,y),(S,S,S))

        Intensity.putpixel((x,y),(I,I,I)) 
    
        #Black tao nen khi gan min(R,G,B) vao ca 3 kenh R,G,B
        HSI.putpixel((x,y),(I,S,H)) 

Hue_CV = np.array(Hue)
Intensity_CV = np.array(Intensity)
Saturation_CV = np.array(Saturation)
HSI_CV = np.array(HSI)

#Hien thi 4 tam hinh tren cung mot cua so
vert=np.concatenate((Hue_CV,Intensity_CV),axis=0)
vert1=np.concatenate((Saturation_CV,HSI),axis=0)
vert2=np.concatenate((vert,vert1),axis=1)

#Dieu chinh kich thuoc cua so hien thi
HSI_program = cv2.resize(vert2,(700,700))

#Hiển thị hình dùng thư viện OpenCV
cv2.imshow('Anh goc', img)
cv2.imshow('HSI',HSI_program)

#Bấm phím bất kì để đóng cửa sổ hiển thị hình
cv2.waitKey(0)

#Giai phong bo nho
cv2.destroyAllWindows()